pkg/statistics/builder.go

```diff
@@ -373,12 +375,53 @@ func BuildHistAndTopN(
 		if err != nil {
 			return nil, nil, errors.Trace(err)
 		}
+		// For debugging invalid sample data.
+		var (
+			foundTwice      bool
+			firstTimeSample types.Datum
+		)
 		for j := 0; j < len(topNList); j++ {
 			if bytes.Equal(sampleBytes, topNList[j].Encoded) {
-				// find the same value in topn: need to skip over this value in samples
+				// This should never happen, but we met this panic before, so we add this check here.
+				// See: https://github.com/pingcap/tidb/issues/35948
+				if foundTwice {
+					datumString, err := samples[i].Value.ToString()
+					if err != nil {
+						logutil.BgLogger().With(
+							zap.String("category", "stats"),
+						).Error("try to convert datum to string failed", zap.Error(err))
+					}
+					firstTimeSampleString, err := firstTimeSample.ToString()
+					if err != nil {
+						logutil.BgLogger().With(
+							zap.String("category", "stats"),
+						).Error("try to convert datum to string failed", zap.Error(err))
+					}
+
+					logutil.BgLogger().With(
+						zap.String("category", "stats"),
+					).Warn(
+						"invalid sample data",
+						zap.Int64("i", i),
+						zap.Int64("j", int64(j)),
+						zap.String("datum", datumString),
+						zap.Binary("sampleBytes", sampleBytes),
+						zap.Binary("topNBytes", topNList[j].Encoded),
+						zap.String("firstTimeSample", firstTimeSampleString),
+					)
```

Comment:
We missed the `hist_id`(col/index's id) and `is_index`.

Reply:
Added.